smmregrid/cdogenerate.py

```diff
@@ -0,0 +1,406 @@
+            Warning: If deprecated arguments `extra` or `vert_coord` are used.
+
+        Notes:
+            This function handles both 2D and 3D grid cases:
+
+            For 2D grids (when `vertical_dim` is None), it calls the `_weights_2d` method
+            to generate weights. The weights are then masked based on a precomputed weights matrix.
+
+            For 3D grids (when `vertical_dim` is specified), it uses multiprocessing to generate weights
+            for each vertical level. It requires the vertical dimension to be present in the source grid,
+            and it will generate a mask indicating valid and invalid weights for each vertical level.
+
+            The function logs the progress of weight generation, including the length of vertical dimensions
+            and each level being processed.
+
+            Deprecation Warning: The `extra` and `vert_coord` parameters are deprecated and will be removed in future versions.
+            Users should migrate to using `cdo_extra` and `vertical_dim`, respectively.
+
+        """
+        if self.target_grid is None:
+            raise TypeError('Target grid is not specified, cannot provide any regridding')
+        if self.source_grid is None:
+            raise TypeError('Source grid is not specified, cannot provide any regridding')
+
+        # verify that method and normalization are suitable
+        self._safe_check(method, remap_norm)
+
+        # vertical dimension
+        vertical_dim = deprecated_argument(vert_coord, vertical_dim, 'vert_coord', 'vertical_dim')
+
+        # Generate weights for 2D or 3D grid based on vertical_dim presence
+        if not vertical_dim:
+            return self._weights_2d(method, extrapolate, remap_norm, remap_area_min)
+        return self._weights_3d(method, extrapolate, remap_norm, remap_area_min,
+                            nproc, vertical_dim)
+
+    def _weights_2d(self, method, extrapolate, remap_norm,
+                remap_area_min):
+        """Generate 2D weights using CDO."""
+
+        weights = self._cdo_generate_weights(method, extrapolate,
+                                             remap_norm, remap_area_min)
+        weights_matrix = compute_weights_matrix(weights)
+        weights = mask_weights(weights, weights_matrix)
+        masked = int(check_mask(weights))
+        masked_xa = xarray.DataArray(masked, name="dst_grid_masked")
+        return xarray.merge([weights, masked_xa])
+
+    def _weights_3d(self, method, extrapolate, remap_norm,
+                remap_area_min, nproc, vertical_dim):
+        """Generate 3D weights using multiprocessing."""
+
+        if isinstance(self.source_grid, str):
+            sgrid = xarray.open_dataset(self.source_grid)
+        else:
+            sgrid = self.source_grid
+
+        if vertical_dim not in sgrid:
+            raise KeyError(f'Cannot find vertical dim {vertical_dim} in {list(sgrid.dims)}')
+
+        nvert = sgrid[vertical_dim].values.size
```

Comment:
This assumes that there is a coordinate `vertical_dim` with the same name as the dimension, correct? Is this robust?

Reply:
@oloapinivad what about this one ?

Reply:
what about `sgrid.sizes[vertical_dim]`. this is cleaner, access metadata and provide the dimension infromation only without mixing up dimension and coordinates...